pkg/translator/zipkin/internal/zipkin/to_domain.go

```diff
@@ -0,0 +1,418 @@
+package zipkin
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/jaegertracing/jaeger-idl/model/v1"
+	"github.com/jaegertracing/jaeger-idl/thrift-gen/zipkincore"
+)
+
+const (
+	// UnknownServiceName is serviceName we give to model.Process if we cannot find it anywhere in a Zipkin span
+	UnknownServiceName = "unknown-service-name"
+	component          = "component"
+	peerservice        = "peer.service"
+	peerHostIPv4       = "peer.ipv4"
+	peerHostIPv6       = "peer.ipv6"
+	peerPort           = "peer.port"
+)
+
+var (
+	coreAnnotations = map[string]string{
+		zipkincore.SERVER_RECV: trace.SpanKindServer.String(),
+		zipkincore.SERVER_SEND: trace.SpanKindServer.String(),
+		zipkincore.CLIENT_RECV: trace.SpanKindClient.String(),
+		zipkincore.CLIENT_SEND: trace.SpanKindClient.String(),
+	}
+
+	// Some tags on Zipkin spans really describe the process emitting them rather than an individual span.
+	// Once all clients are upgraded to use native Jaeger model, this won't be happenning, but for now
+	// we remove these tags from the span and store them in the Process.
+	processTagAnnotations = map[string]string{
+		"jaegerClient":    "jaeger.version", // transform this tag name to client.version
+		"jaeger.hostname": "hostname",       // transform this tag name to hostname
+		"jaeger.version":  "jaeger.version", // keep this key as is
+	}
+
+	trueByteSlice = []byte{1}
+
+	// DefaultLogFieldKey is the log field key which translates directly into Zipkin's Annotation.Value,
+	// provided it's the only field in the log.
+	// In all other cases the fields are encoded into Annotation.Value as JSON string.
+	// TODO move to domain model
+	DefaultLogFieldKey = "event"
+
+	// IPTagName is the Jaeger tag name for an IPv4/IPv6 IP address.
+	// TODO move to domain model
+	IPTagName = "ip"
+)
+
+// ToDomain transforms a trace in zipkin.thrift format into model.Trace.
+// The transformation assumes that all spans have the same Trace ID.
+// A valid model.Trace is always returned, even when there are errors.
+// The errors are more of an "fyi", describing issues in the data.
+// TODO consider using different return type instead of `error`.
+func ToDomain(zSpans []*zipkincore.Span) (*model.Trace, error) {
```

Comment:
once this is merged may want to book another ticket to implement direct translation zipkin->OTLP, there's no reason why this needs to go through the jaeger model, just an extra overhead.